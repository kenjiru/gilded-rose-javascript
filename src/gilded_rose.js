export const ItemType = {
    LEGENDARY: 'Sulfuras',
    AGED_BRIE: 'Aged Brie',
    TICKET: 'Backstage passes',
    CONJURED: 'Conjured'
};

export class Item {
    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

const MIN_QUALITY = 0;
const MAX_QUALITY = 50;

const QUALITY_INCREMENT = 1;
const SELL_IN_INCREMENT = 1;

const SELL_IN_VALUE = 0;
const TICKET_FIRST_SELL_OUT = 10;
const TICKET_SECOND_SELL_OUT = 5;

export class Shop {
    constructor(items = []) {
        this.items = items;
    }

    updateQuality() {
        for (var i = 0; i < this.items.length; i++) {
            const item = this.items[i];

            this.handleQuality(item);
            this.handleSellIn(item);
            this.handleQualityAfterSellIn(item);
        }

        return this.items;
    }

    handleSellIn(item) {
        if (Shop.isItemOfType(item, ItemType.LEGENDARY) === false) {
            item.sellIn = item.sellIn - SELL_IN_INCREMENT;
        }
    }

    handleQuality(item) {
        if (Shop.isItemOfType(item, ItemType.AGED_BRIE) || Shop.isItemOfType(item, ItemType.TICKET)) {
            this.handleQualityForBrieAndTicket(item);
            return;
        }

        if (Shop.isItemOfType(item, ItemType.LEGENDARY) === false) {
            this.decrementQuality(item);
        }
    }

    handleQualityForBrieAndTicket(item) {
        if (Shop.isItemOfType(item, ItemType.AGED_BRIE)) {
            this.incrementQuality(item);
            return;
        }

        if (Shop.isItemOfType(item, ItemType.TICKET)) {
            this.handleTicketQuality(item);
        }
    }

    handleTicketQuality(item) {
        this.incrementQuality(item);

        if (item.sellIn <= TICKET_FIRST_SELL_OUT) {
            this.incrementQuality(item);
        }

        if (item.sellIn <= TICKET_SECOND_SELL_OUT) {
            this.incrementQuality(item);
        }
    }

    handleQualityAfterSellIn(item) {
        if (item.sellIn >= SELL_IN_VALUE) {
            return;
        }

        if (Shop.isItemOfType(item, ItemType.AGED_BRIE)) {
            this.incrementQuality(item);
            return;
        }

        this.handleQualityForNonAgingItems(item);
    }

    handleQualityForNonAgingItems(item) {
        if (Shop.isItemOfType(item, ItemType.TICKET)) {
            item.quality = MIN_QUALITY;
            return;
        }

        if (Shop.isItemOfType(item, ItemType.LEGENDARY) === false) {
            this.decrementQuality(item);
        }
    }

    incrementQuality(item) {
        if (item.quality < MAX_QUALITY) {
            item.quality = item.quality + QUALITY_INCREMENT;
        }
    }

    decrementQuality(item) {
        if (item.quality > MIN_QUALITY) {
            item.quality = item.quality - QUALITY_INCREMENT;
        }
    }

    getFirstItemOfType(itemType) {
        return Shop.getFirstItemOfGivenType(this.items, itemType);
    }

    static getFirstItemOfGivenType(items, itemType) {
        return items.find((item) => {
            return Shop.isItemOfType(item, itemType);
        })
    }

    static isItemOfType(item, itemType) {
        return item.name.startsWith(itemType);
    }
}
