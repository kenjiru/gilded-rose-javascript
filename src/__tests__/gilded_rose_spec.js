import {Shop, Item} from '../gilded_rose';

const createFixtureShop = () => {
    const items = [];

    items.push(new Item('+5 Dexterity Vest', 10, 20));
    items.push(new Item('Aged Brie', 2, 0));
    items.push(new Item('Aged Brie', 2, 49));
    items.push(new Item('Elixir of the Mongoose', 5, 7));
    items.push(new Item('Sulfuras, Hand of Ragnaros', 0, 80));
    items.push(new Item('Sulfuras, Hand of Ragnaros', -1, 80));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 10, 49));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 10, 45));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49));
    items.push(new Item('Backstage passes to a TAFKAL80ETC concert', 5, 45));
    // this conjured item does not work properly yet
    items.push(new Item('Conjured Mana Cake', 3, 6));


    return new Shop(items);
};

const cloneDeep = function (data) {
    return JSON.parse(JSON.stringify(data));
};

describe("Gilded Rose", () => {
    it("works as the original", () => {
        const shop = createFixtureShop();
        const days = 10;
        const actual = [];

        for (let i = 0; i < days; i++) {
            actual.push(cloneDeep(shop.items));
            shop.updateQuality();
        }

        expect(actual).toMatchSnapshot();
    });
});