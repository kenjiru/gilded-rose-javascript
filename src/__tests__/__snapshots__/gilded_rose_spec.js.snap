// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose works as the original 1`] = `
Array [
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 20,
      "sellIn": 10,
    },
    Object {
      "name": "Aged Brie",
      "quality": 0,
      "sellIn": 2,
    },
    Object {
      "name": "Aged Brie",
      "quality": 49,
      "sellIn": 2,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 7,
      "sellIn": 5,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 20,
      "sellIn": 15,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 49,
      "sellIn": 10,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 45,
      "sellIn": 10,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 49,
      "sellIn": 5,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 45,
      "sellIn": 5,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 6,
      "sellIn": 3,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 19,
      "sellIn": 9,
    },
    Object {
      "name": "Aged Brie",
      "quality": 1,
      "sellIn": 1,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": 1,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 6,
      "sellIn": 4,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 21,
      "sellIn": 14,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 9,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 47,
      "sellIn": 9,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 4,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 48,
      "sellIn": 4,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 4,
      "sellIn": 2,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 18,
      "sellIn": 8,
    },
    Object {
      "name": "Aged Brie",
      "quality": 2,
      "sellIn": 0,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": 0,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 5,
      "sellIn": 3,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 22,
      "sellIn": 13,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 8,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 49,
      "sellIn": 8,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 3,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 3,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 2,
      "sellIn": 1,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 17,
      "sellIn": 7,
    },
    Object {
      "name": "Aged Brie",
      "quality": 4,
      "sellIn": -1,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -1,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 4,
      "sellIn": 2,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 23,
      "sellIn": 12,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 7,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 7,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 2,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 2,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": 0,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 16,
      "sellIn": 6,
    },
    Object {
      "name": "Aged Brie",
      "quality": 6,
      "sellIn": -2,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -2,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 3,
      "sellIn": 1,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 24,
      "sellIn": 11,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 6,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 6,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 1,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": -1,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 15,
      "sellIn": 5,
    },
    Object {
      "name": "Aged Brie",
      "quality": 8,
      "sellIn": -3,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -3,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 2,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 25,
      "sellIn": 10,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 5,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 5,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 0,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 0,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": -2,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 14,
      "sellIn": 4,
    },
    Object {
      "name": "Aged Brie",
      "quality": 10,
      "sellIn": -4,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -4,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 0,
      "sellIn": -1,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 27,
      "sellIn": 9,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 4,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 4,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -1,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": -3,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 13,
      "sellIn": 3,
    },
    Object {
      "name": "Aged Brie",
      "quality": 12,
      "sellIn": -5,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -5,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 0,
      "sellIn": -2,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 29,
      "sellIn": 8,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 3,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 3,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -2,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -2,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": -4,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 12,
      "sellIn": 2,
    },
    Object {
      "name": "Aged Brie",
      "quality": 14,
      "sellIn": -6,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -6,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 0,
      "sellIn": -3,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 31,
      "sellIn": 7,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 2,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 2,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -3,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -3,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": -5,
    },
  ],
  Array [
    Object {
      "name": "+5 Dexterity Vest",
      "quality": 11,
      "sellIn": 1,
    },
    Object {
      "name": "Aged Brie",
      "quality": 16,
      "sellIn": -7,
    },
    Object {
      "name": "Aged Brie",
      "quality": 50,
      "sellIn": -7,
    },
    Object {
      "name": "Elixir of the Mongoose",
      "quality": 0,
      "sellIn": -4,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": 0,
    },
    Object {
      "name": "Sulfuras, Hand of Ragnaros",
      "quality": 80,
      "sellIn": -1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 33,
      "sellIn": 6,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 50,
      "sellIn": 1,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -4,
    },
    Object {
      "name": "Backstage passes to a TAFKAL80ETC concert",
      "quality": 0,
      "sellIn": -4,
    },
    Object {
      "name": "Conjured Mana Cake",
      "quality": 0,
      "sellIn": -6,
    },
  ],
]
`;
